libarchive/test/test_write_format_zip_compression_lzmaxz.c

```diff
@@ -396,5 +405,63 @@ DEFINE_TEST(test_write_format_zip_compression_lzma)
 	dumpfile("constructed.zip", buff, used);
 
 	verify_lzma_contents(buff, used);
+
+	/* Same song and dance, but for XZ */
+
+	/* Create new ZIP archive in memory without padding. */
+	/* Use the setter function to use XZ compression. */
+	assert((a = archive_write_new()) != NULL);
+	assertEqualIntA(a, ARCHIVE_OK, archive_write_set_format_zip(a));
+	assertEqualIntA(a, ARCHIVE_OK, archive_write_zip_set_compression_xz(a));
+	assertEqualIntA(a, ARCHIVE_OK,
+	    archive_write_set_options(a, "zip:experimental"));
+	assertEqualIntA(a, ARCHIVE_OK, archive_write_add_filter_none(a));
+	assertEqualIntA(a, ARCHIVE_OK, archive_write_set_bytes_per_block(a, 1));
+	assertEqualIntA(a, ARCHIVE_OK, archive_write_set_bytes_in_last_block(a, 1));
+	assertEqualIntA(a, ARCHIVE_OK, archive_write_open_memory(a, buff, sizeof(buff), &used));
+
+	verify_write_lzma(a);
+
+	/* Close the archive . */
+	assertEqualIntA(a, ARCHIVE_OK, archive_write_close(a));
+	assertEqualInt(ARCHIVE_OK, archive_write_free(a));
+	dumpfile("constructed.zip", buff, used);
+
+	verify_xz_contents(buff, used, 76);
+
+	/* Create new ZIP archive in memory without padding. */
+	/* Use compression-level=9 to check that compression levels are
+	 * somewhat supported as well as threads=2 to check the multi-threaded
+	 * encoder, if available. */
+	assert((a = archive_write_new()) != NULL);
+	assertEqualIntA(a, ARCHIVE_OK, archive_write_set_format_zip(a));
+	assertEqualIntA(a, ARCHIVE_OK,
+	    archive_write_set_options(a, "zip:compression=xz"));
+	assertEqualIntA(a, ARCHIVE_OK,
+	    archive_write_set_options(a, "zip:compression-level=9"));
+	assertEqualIntA(a, ARCHIVE_OK,
+	    archive_write_set_options(a, "zip:threads=2"));
+	assertEqualIntA(a, ARCHIVE_OK,
+	    archive_write_set_options(a, "zip:experimental"));
+	assertEqualIntA(a, ARCHIVE_OK, archive_write_add_filter_none(a));
+	assertEqualIntA(a, ARCHIVE_OK, archive_write_set_bytes_per_block(a, 1));
+	assertEqualIntA(a, ARCHIVE_OK, archive_write_set_bytes_in_last_block(a, 1));
+	assertEqualIntA(a, ARCHIVE_OK, archive_write_open_memory(a, buff, sizeof(buff), &used));
+
+	verify_write_lzma(a);
+
+	/* Close the archive. */
+	assertEqualIntA(a, ARCHIVE_OK, archive_write_close(a));
+	assertEqualInt(ARCHIVE_OK, archive_write_free(a));
+	dumpfile("constructed.zip", buff, used);
+
+/* The compressed size changes depending on if we have the multi-threaded
+ * encoder at hand or not, given that we are asking for it to be used if
+ * possible */
+#ifdef HAVE_LZMA_STREAM_ENCODER_MT
+	verify_xz_contents(buff, used, 84);
+#else
+	verify_xz_contents(buff, used, 76);
```

Comment:
I'll have to see if it could be that but it's possible that it's due to the non-threaded encoder never storing the compressed and uncompressed size of LZMA blocks while the threaded encoder always does, because the block sizes are mandatory for threaded decompression.

Reply:
After checking it isn't even due to the block sizes: they would have fit in the single-threaded's block headed padding. It's due to, for reasons I haven't been able to find, the multi-threaded encoded adding 8 bytes of padding over the, if I didn't misread the [XZ spec](https://tukaani.org/xz/xz-file-format.txt), the necessary 1 byte it still needed to put in. The one saving grace is that, per [xz's man page](https://git.tukaani.org/?p=xz.git;a=blob_plain;f=src/xz/xz.1;hb=HEAD), it's expected that the multi-threaded encoder produce bigger files than the mono-threaded one.

Reply:
Hard-coding the size here is going to be brittle.  Many compression libraries evolve over time in ways that cause the actual size for any particular piece of data to change.  Also, I'm not at all surprised that the MT encoder produces slightly larger results than the single-threaded encoder; that's quite common.

Reply:
Then maybe I should skip checking the compressed size and assume that as long as the round-trip is successful, it's good?

Reply:
That's what I would do, yes.

Reply:
I changed the tests accordingly, now they're only checking that the round-trip passed using libarchive's reader.